---
title: "Analysis Colon dataset - GSE132465"
author: "Lena Cords"
output:
  html_document:
    df_print: paged
---



```{r load libraries}
suppressMessages({
library(splatter)
library(Seurat)
library(msigdbr)
library(singleseqgset)
library(heatmap3)
library(GEOquery)
library(Seurat) 
library(scater)
library(SingleCellExperiment)
library(viridisLite)
library(viridis)
library(clustree)
library(MAST)
library(destiny)
library(ggbeeswarm)
library(ggthemes)
library(dplyr)
library(tidyr)
library(Seurat)
library(stringr)
library(colorRamps)
library(colorRamps)
library(pheatmap)
library(RColorBrewer)
library(dplyr)
library(singleseqgset)
library(devtools)
})
```

```{r}
set.seed(101100)
```

Set working directory
```{r}
wd <- dirname(getwd())
plot.folder <- file.path(wd,"plots")
data.folder <- file.path(wd,"data","colon")
```

```{r, eval=FALSE}
    #GSE132465
dat_gse_132465<-getGEOSuppFiles('GSE132465')
####
#temp <- read.table(gzfile("~/lena_Raw/BC_Fibro/scRNA/Public data/Colon/R/GSE132465/GSE132465_GEO_processed_CRC_10X_raw_UMI_count_matrix.txt.gz"), sep="\t", header = T)

temp <- read.table(gzfile("~/data/lena_Raw/colon/GSE132465/GSE132465_GEO_processed_CRC_10X_raw_UMI_count_matrix.txt.gz"), sep="\t", header = T)

temp2 <- read.table(gzfile("~/data/lena_Raw/colon/GSE132465/GSE132465_GEO_processed_CRC_10X_natural_log_TPM_matrix.txt.gz"), sep="\t", header = T)

temp3 <- read.table(gzfile("~/data/lena_Raw/colon/GSE132465/GSE132465_GEO_processed_CRC_10X_cell_annotation.txt.gz"), sep="\t", header = T)

head(temp2)
head(temp3)

colnames(temp2)<-gsub('\\.', '-', colnames(temp2))
df_m <-temp2
#head(df_m)
df_m <- as.matrix(df_m[,-1])
rownames(df_m)<-temp2$Index

#create Seurat Object with metadata
#Colon.Seurat <- CreateSeuratObject(counts=df_m,meta.data = meta_data_complete_gse_GSE132465)
rownames(temp3)<-temp3$Index

wd <- dirname(getwd())%>% dirname

Colon.Seurat <- CreateSeuratObject(counts=df_m,meta.data = temp3)
saveRDS(Colon.Seurat, file=file.path(wd,"Breast_scRNA","data","colon_RAW.rds" ))
saveRDS(Colon.Seurat, file=file.path(data.folder, paste("Colon_Seurat.RDS")))
```

Load data
```{r}
Colon.Seurat <-readRDS(file=file.path(data.folder, "Colon-Seurat_workingfile_new.rds"))
```

```{r, eval=FALSE}
Colon.Seurat[["percent.mt"]] <- PercentageFeatureSet(Colon.Seurat, pattern = "^MT-")
Colon.Seurat <- PercentageFeatureSet(Colon.Seurat, pattern = "^KRT", col.name = "percent.krt")
Colon.Seurat <- PercentageFeatureSet(Colon.Seurat, pattern = "MGP", col.name = "percent.MGP")

Colon.Seurat <- SCTransform(Colon.Seurat , vars.to.regress = c("percent.mt", "percent.krt", "percent.MGP"), verbose = TRUE)
VariableFeatures(Colon.Seurat[["SCT"]]) <- rownames(Colon.Seurat[["SCT"]]@scale.data)

VlnPlot(Colon.Seurat, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","percent.MGP","percent.krt"), ncol = 5)

FindVariableFeatures(Colon.Seurat)
```

```{r, fig.width=10}
plot1 <- FeatureScatter(Colon.Seurat, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Colon.Seurat, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r, eval=FALSE}
#r fewer than 201 UMIs, over 6,000 or below 101 expressed genes, or over 10% UMIs derived from mitochondrial genome.
Colon.Seurat <- subset(Colon.Seurat, subset = nFeature_RNA > 200 & nFeature_RNA < 6000 & percent.mt < 10)
```

```{r, eval=FALSE}
saveRDS(Colon.Seurat, file=file.path(data.folder,"colon", paste("Colon-Seurat_workingfile.rds")))
Colon.Seurat <-readRDS(file=file.path(data.folder,"colon", paste("Colon-Seurat_workingfile.rds")))
```

```{r, eval=FALSE}
Colon.Seurat <- RunPCA(object = Colon.Seurat)
ElbowPlot(Colon.Seurat)

print(x = Colon.Seurat[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)
PCA_genes <- VizDimLoadings(object = Colon.Seurat, dims = 1:2)

DimHeatmap(object = Colon.Seurat, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(object = Colon.Seurat, dims = 10:20, cells = 500, balanced = TRUE)
ElbowPlot(object = Colon.Seurat, ndims =40)


Colon.Seurat <- RunUMAP(object = Colon.Seurat, dims = 1:20)
#Colon.Seurat <- RunTSNE(object = Colon.Seurat, dims = 1:20)

DimPlot(object = Colon.Seurat, reduction = 'umap',group.by = "Cell_type", label = F)
saveRDS(Colon.Seurat, file=file.path(data.folder, paste("Colon-Seurat_workingfile_new.rds")))
```
```{r}
DimPlot(object = Colon.Seurat, reduction = 'umap',group.by = "Cell_type", label = F)
```

```{r, eval=F}
#set dimensions to be used
Colon.Seurat <- FindNeighbors(object = Colon.Seurat, dims = 1:20,graph.name = "all.colon")
#set resolution for clustering ( this has to be changed in otder to determine the best clustering resolution)
Colon.Seurat <- FindClusters(object = Colon.Seurat, resolution = seq(from = 0.1, to = 2.5, by = 0.1), graph.name = "all.colon")
```

```{r, fig.width=12, fig.height=12}
p<-clustree(Colon.Seurat, prefix = "all.colon_res.0.", exprs = "scale.data")
plot(p)
#ggsave(filename=file.path(plot.folder, paste("clustree_all0-subs.pdf")), plot=p, width=10, height=10)
```

```{r, fig.width=12, fig.height=8}
p <-FeaturePlot(Colon.Seurat, features=c("ACTA2","RGS5","MCAM","CD36","MYH11","FAP","PDPN","CFD","PLA2G2A","MME","MMP11","CCL19","CD74"),reduction="umap",cols=c("lightgrey","black"), ncol=4) &NoAxes()
plot(p)

p<-
  FeaturePlot(Colon.Seurat, features=c("ACTA2","PECAM1","PTPRC","PDPN","MCAM","CD3E","MS4A1","KRT7","KRT19","HLA-DRB1"), cols=c("lightgrey","black"),ncol=5)  &NoAxes()
p

#ggsave(filename=file.path(plot.folder, paste("UMAP_FeaturePlot_Colon_sub.png")), plot=p, width=12, height=6, dpi=300)
```

```{r, eval=F}
Idents(Colon.Seurat) <- Colon.Seurat$all.colon_res.0.2

markers.clean.MAST_Colon_02<- FindAllMarkers(object = Colon.Seurat, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, test.use="MAST")
write.csv(markers.clean.MAST_Colon_02, file = file.path(data.folder,paste( "DE_cluster_AllMarkerGenes_MAST_Colon_res02_new2.csv")), row.names = FALSE)

markers.clean.MAST_Colon_02_top20 <- markers.clean.MAST_Colon_02 %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
#write.csv(markers.clean.MAST_Colon.sub_02_top20, file = file.path(data.folder,paste0( "DE_cluster_AllMarkerGenes_MAST_Colon.sub_res02_top20_new.csv")), row.names = TRUE)
DimPlot(object = Colon.Seurat, reduction = 'umap',group.by = "Cell_type", label = F)
markers.clean.MAST_Colon_02_top20

DoHeatmap(Colon.Seurat, features = markers.clean.MAST_Colon_02_top20$gene) + NoLegend()+scale_fill_viridis_c()

```
```{r, fig.width=15, fig.height=15}
markers.clean.MAST_Colon_02 <-read.csv(file = file.path(data.folder,paste( "DE_cluster_AllMarkerGenes_MAST_Colon_res02_new2.csv")))

markers.clean.MAST_Colon_02_top20 <- markers.clean.MAST_Colon_02 %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)

DoHeatmap(subset(Colon.Seurat, downsample = 10000), features = markers.clean.MAST_Colon_02_top20$gene)+ NoLegend()+scale_fill_viridis_c()

```

```{r}
DimPlot(object = Colon.Seurat, reduction = 'umap',group.by = "all.colon_res.0.2", label = T)
```

```{r}
markers.clean.MAST_Colon_02_top20
```

```{r, eval=F}
colon.fibro.me <- subset(x = Colon.Seurat, subset = all.colon_res.0.2 == "6")

saveRDS(colon.fibro.me, file=file.path(data.folder, paste("colon_fibro_me_RAW2.rds")))

```

Load CAF data
```{r}
colon.fibro.me <- readRDS(file=file.path(data.folder, "colon_fibro_me_wf_clusterd.rds"))
```

#my annotation of CAFs

```{r, eval=F}
colon.fibro.me <- RunPCA(object = colon.fibro.me)
ElbowPlot(colon.fibro.me)

print(x = colon.fibro.me[['pca']], dims = 1:5, nfeatures = 5, projected = FALSE)
PCA_genes <- VizDimLoadings(object = colon.fibro.me, dims = 1:2)

DimHeatmap(object = colon.fibro.me, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(object = colon.fibro.me, dims = 10:20, cells = 500, balanced = TRUE)
ElbowPlot(object = colon.fibro.me, ndims =40)


colon.fibro.me <- RunUMAP(object = colon.fibro.me, dims = 1:20)
#colon.fibro.me <- RunTSNE(object = colon.fibro.me, dims = 1:20)

DimPlot(object = colon.fibro.me, reduction = 'umap',group.by = "Class", label = F,)
saveRDS(colon.fibro.me, file=file.path(data.folder,"colon", paste("colon_fibro_me_wf.rds")))
saveRDS(colon.fibro.me, file=file.path(data.folder, paste("colon_fibro_me_wf.rds")))
colon.fibro.me <- readRDS(file=file.path(data.folder, paste("colon_fibro_me_wf.rds")))

```

```{r}
DimPlot(object = colon.fibro.me, reduction = 'umap',group.by = "Class", label = F)+theme_void()
```

```{r, fig.width=12, fig.height=10}
my_genes <-c("FAP","PDPN","ACTA2","PLA2G2A","CD34","CFD","MMP11","COL1A1","POSTN","MME","NDRG1","ENO1","RGS5","MCAM","CCL21","CL19","HLA-DR","CD74","MKI67","IDO1","HSPH1")

p <-FeaturePlot(colon.fibro.me, features=my_genes,reduction="umap",cols=c("lightgrey","black"), ncol=4) &NoAxes()
plot(p)
#ggsave(filename=file.path(plot.folder, paste("UMAP_FeaturePlot_Colon_sub.png")), plot=p, width=12, height=6, dpi=300)
```

```{r, eval=FALSE}
#set dimensions to be used
colon.fibro.me <- FindNeighbors(object = colon.fibro.me, dims = 1:20,graph.name = "all.subs")
#set resolution for clustering ( this has to be changed in otder to determine the best clustering resolution)
colon.fibro.me <- FindClusters(object = colon.fibro.me, resolution = seq(from = 0.1, to = 1.5, by = 0.1), graph.name = "all.subs")
saveRDS(colon.fibro.me, file=file.path(data.folder, paste("colon_fibro_me_wf.rds")))
```


```{r, fig.width=12, fig.height=12}
p<-clustree(colon.fibro.me, prefix = "all.subs_res.0.", exprs = "scale.data")
plot(p)
```



```{r, eval=FALSE}
Idents(colon.fibro.me) <- colon.fibro.me$all.subs_res.0.3

markers.clean.MAST_colon.fibro.me_03<- FindAllMarkers(object = colon.fibro.me, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, test.use="MAST")
write.csv(markers.clean.MAST_colon.fibro.me_03, file = file.path(data.folder,paste( "DE_cluster_AllMarkerGenes_MAST_colon.fibro.me_res03_new2.csv")), row.names = FALSE)

markers.clean.MAST_colon.fibro.me_03_top20 <- markers.clean.MAST_colon.fibro.me_03 %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
#write.csv(markers.clean.MAST_colon.fibro.me_03_top20, file = file.path(data.folder,paste0( "DE_cluster_AllMarkerGenes_MAST_colon.fibro.me_res03_top20_new.csv")), row.names = TRUE)

DoHeatmap(colon.fibro.me, features = markers.clean.MAST_colon.fibro.me_03_top20$gene) + NoLegend()+scale_fill_viridis_c()
```

```{r}
DimPlot(object = colon.fibro.me, reduction = 'umap',group.by = "all.subs_res.0.3", label = T)
```

```{r}
markers.clean.MAST_colon.fibro.me_03 <- read.csv(file = file.path(data.folder,paste( "DE_cluster_AllMarkerGenes_MAST_colon.fibro.me_res03_new2.csv")))

markers.clean.MAST_colon.fibro.me_03_top20 <- markers.clean.MAST_colon.fibro.me_03 %>% group_by(cluster) %>% top_n(n = 20, wt = avg_log2FC)
DoHeatmap(colon.fibro.me, features = markers.clean.MAST_colon.fibro.me_03_top20$gene) + NoLegend()+scale_fill_viridis_c()
```

```{r}
markers.clean.MAST_colon.fibro.me_03_top20
```


```{r, eval=FALSE}
Idents(colon.fibro.me) <- colon.fibro.me$all.subs_res.0.3

#add sCAFs / mCAFs to labels
x <- c(0:9)
y <-c("iCAF","mCAF","iCAF","mCAF","Pericyte","mCAF","other","tpCAF","tpCAF","rCAF")
cluster_names <- as.data.frame(cbind(x,y))
colnames(cluster_names) <-c("Cluster","CAF_type")

cluster_ft <- cluster_names$CAF_type
cluster_ft <-as.character(cluster_ft)
names(x = cluster_ft) <- levels(x = colon.fibro.me)
colon.fibro.me <- RenameIdents(object = colon.fibro.me, cluster_ft)

#Save celltype information in metadata
colon.fibro.me[["cluster_ft"]] <- Idents(object = colon.fibro.me)
```

```{r}
x <- c(0:9)
y <-c("iCAF","mCAF","iCAF","mCAF","Pericyte","mCAF","other","tpCAF","tpCAF","rCAF")
cluster_names <- as.data.frame(cbind(x,y))
colnames(cluster_names) <-c("Cluster","CAF_type")
cluster_names
```

Top differentially expressed markers per CAF type
```{reval=FALSE}
Idents(colon.fibro.me) <- colon.fibro.me$cluster_ft

markers.clean.MAST_Colon.fibro_cluster<- FindAllMarkers(object = colon.fibro.me, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25, test.use="MAST")
#
write.csv(markers.clean.MAST_Colon.fibro_cluster, file = file.path(data_folder,"colon",paste( "DE_cluster_AllMarkerGenes_MAST_Colon.fibro_res-cluster.csv")), row.names = FALSE)

markers.clean.MAST_Colon.fibro_cluster_top10 <- markers.clean.MAST_Colon.fibro_cluster %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
write.csv(markers.clean.MAST_Colon.fibro_cluster_top10, file = file.path(data_folder,"colon",paste0( "DE_cluster_AllMarkerGenes_MAST_Colon.fibro_res-cluster_top10.csv")), row.names = TRUE)
```


```{r, fig.height=6, fig.width=12}
Idents(colon.fibro.me) <- colon.fibro.me$cluster_ft

markers.clean.MAST_Colon.fibro_cluster <- read.csv(file = file.path(data.folder,paste( "DE_cluster_AllMarkerGenes_MAST_Colon.fibro_res-cluster.csv")))
markers.clean.MAST_Colon.fibro_cluster_top10 <- markers.clean.MAST_Colon.fibro_cluster %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)

p <-DoHeatmap(colon.fibro.me, features = markers.clean.MAST_Colon.fibro_cluster_top10$gene) +scale_fill_viridis_c()
p
#ggsave(p, file=file.path(plot.folder, paste("Colon_fibros_cluster_Heatmap.png")))
```


CAF type distribution over tumour and normal tissue
```{r}
df <- prop.table(x=table(colon.fibro.me$Class, colon.fibro.me$cluster_ft), margin=2)
df <- as.data.frame(df)
colnames(df) <- c('Dataset', 'Cluster', 'Prop')

p <-
  ggplot(df, aes(x=Cluster, y = Prop, fill=Dataset)) +
  #scale_fill_viridis(discrete=TRUE)+
  geom_bar(stat="identity", position="fill")+
  theme(axis.title.x=element_blank(), axis.title.y = element_blank())+
  coord_flip()+
  theme(panel.background = element_blank(), axis.text.x = element_text(hjust=1, angle=45))+
  ggtitle('CAF phenotype distribution per patient')+
  scale_fill_tableau()
p
#ggsave(plot=p, file=file.path(data_folder,"colon", paste("BARplot_Colon_CAFs over tissue.pdf")))
```



```{r}
Idents(colon.fibro.me) <- colon.fibro.me$cluster_ft
#umap
p <-DimPlot(object = colon.fibro.me, reduction = 'umap',group.by = "cluster_ft", label = F)+labs(colour="cluster_ft")+theme_void()
plot(p)
#ggsave(p, file=file.path(data_folder, "colon","UMAP_CAFtype.png"), dpi=300)
```



```{r, eval=FALSE}
saveRDS(colon.fibro.me, file=file.path(data.folder, paste("colon_fibro_me_wf_clusterd.rds")))
```

Subset to tumour only
```{r}
colon.fibro.me.tumour <- subset(x = colon.fibro.me, subset = Class == "Tumor" &
                                 cluster_ft !="other")
```

```{r, eval=FALSE}
colon.fibro.me.tumour <- subset(x = colon.fibro.me, subset = Class == "Tumor" &
                                 cluster_ft !="other")

saveRDS(colon.fibro.me.tumour, file=file.path(data.folder,"COLON_tumour_fibros_clustered_me_minusOther.rds"))
```

Gene set enrichment analysis
```{r}
#Hallmark genes
h.human <- msigdbr(species="Homo sapiens",category="H")

h.names <- unique(h.human$gs_name)

h.sets <- vector("list",length=length(h.names))
names(h.sets) <- h.names

for (i in names(h.sets)) {
    h.sets[[i]] <- pull(h.human[h.human$gs_name==i,"gene_symbol"])
}
```

```{r}
colon.fibro.me.tumour$cluster_ft <- droplevels(colon.fibro.me.tumour$cluster_ft)
logfc.data <- singleseqgset::logFC(cluster.ids=colon.fibro.me.tumour@meta.data$cluster_ft,
                    expr.mat=colon.fibro.me.tumour[["SCT"]]@scale.data)
```

```{r}
gse.res <- wmw_gsea(expr.mat=colon.fibro.me.tumour[["SCT"]]@scale.data,cluster.cells=logfc.data[[1]],log.fc.cluster=logfc.data[[2]],gene.sets=h.sets)
```

```{r}
res.stats <- gse.res[["GSEA_statistics"]]
res.pvals <- gse.res[["GSEA_p_values"]]

res.pvals <- apply(res.pvals,2,p.adjust,method="fdr") #Correct for multiple comparisons

res.stats[order(res.stats[,1],decreasing=TRUE)[1:10],] #Top gene sets enriched by z scores
```

```{r}
res.pvals[order(res.stats[,1],decreasing=TRUE)[1:10],] #Top gene sets by p values
```


```{r, fig.width=8, fig.height=18}

df <- res.stats
df$hallmark <- rownames(df)
df <- df %>%
      mutate_at("hallmark", str_replace, "HALLMARK_", "")
rownames(df) <- df$hallmark
df$hallmark <-NULL


#pheatmap(as.matrix(df),   color = blue2red(200))

pheatmap(as.matrix(res.stats),   color = blue2red(200))

rdylbu <-rev(brewer.pal(11,"RdYlBu"))


pheatmap(as.matrix(df),  breaks = seq(-4, 4, length.out = 200),   color = colorRampPalette(c("navy", "white", "red"))(200))

#pdf(file=file.path(data_folder, "colon","GSEA_CAFtypes_n_validation.pdf"), height=8, width=18)
#pheatmap(t(as.matrix(df)),  breaks = seq(-4, 4, length.out = 200),   color = colorRampPalette(c("navy", "white", "red"))(200))
#dev.off()
```